### PR TITLE
editoast: add get_catenaries_on_path endpoint

### LINKED ISSUE #3605: Accessing electrification on the path from the front
## Pitch

Add in the pathfinding's result the different mode track ranges found on the path.

## Acceptance criteria

- [ ] The pathfinding's result contains the electrification modes

Image

### PR DESCRIPTION
**Pull request by Castavo**
closes: #3605 

In this PR, I add:
* Methods on the `Pathfinding ` model.
* A `Catenary` model specifically designed for fixtures.
* A `catenaries_on_pathfinding` view.
* An `empty_infra` fixture in `fixture.rs` and a `infra_with_catenaries` fixture.

### CODE REVIEW

File: editoast/src/models/pathfinding.rs
Comment (leovalais) [inline editoast/src/models/pathfinding.rs:147]: That's nice! But why not turn it into a fixture?

File: editoast/src/views/pathfinding/catenaries.rs
Comment (leovalais) [inline editoast/src/views/pathfinding/catenaries.rs:52]: Why not derive(Eq)?

**Review by leovalais — DISMISSED** [review]
LGTM :)

File: editoast/src/models/pathfinding.rs
Comment (Castavo) [reply editoast/src/models/pathfinding.rs:147]: I could add a wrapper around this function to call `create` on the pathfinding and wrap it in `TestFixture` but I would still need to pass `infra_id` as an arg to the fixture, and from what I understood from the rstest documentation, it's not possible (I might have misunderstood)

File: editoast/src/views/pathfinding/catenaries.rs
Comment (Castavo) [reply editoast/src/views/pathfinding/catenaries.rs:52]: Because `f64` is not `Eq` itself, saddly, so you have to do it that way according to the [docs](https://doc.rust-lang.org/std/cmp/trait.Eq.html#derivable)

File: editoast/src/infra.rs
Comment (flomonster) [inline editoast/src/infra.rs]: This is done and will conflict with #3718.
Are you okay to rebase and merge this PR after that?

File: editoast/src/models/pathfinding.rs
Comment (flomonster) [inline editoast/src/models/pathfinding.rs:1]: We can add documentation to the new public functions.

File: editoast/src/models/pathfinding.rs
Comment (flomonster) [reply editoast/src/models/pathfinding.rs:147]: You can do this but not for a given infra. 

Two options:

- We can have a fixture that creates both an infra and a path.
- We can ignore the infra_id parameter and set it to `0` since it's not used in tests

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs:68]: In my experience the fastest way to dedup a collection in rust is not using an hashset.
You collect into a `Vec`. Then `Vec`, then sort it and call the [dedup](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.dedup) function.

Here is an example:

https://github.com/flomonster/kdtree-ray/blob/afdf6562b5a72823eacbec4dd415fb3da698bed7/src/kdtree.rs#L58-L87

Notes:
- Doing this might be overkill here (You can ignore this comment if the endpoint response takes less than a few milliseconds) 
- Otherwise, it can be interesting to benchmark this.

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]:
```suggestion
            if track_ids.contains(track_range.track.as_str()) {
```

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: We don't generate catenaries error yet. I don't know if overlapping catenary should be an error or warning. In all cases, an infra can have overlapping catenary and we should not panic.

We can return an error for now.

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs:256]: Macro master ! :1st_place_medal:

**Review by flomonster — CHANGES_REQUESTED** [review]
Great PR! Some changes will have to be made once the #3718 is merged.

File: editoast/src/models/pathfinding.rs
Comment (Castavo) [reply editoast/src/models/pathfinding.rs:147]: It's not used in tests but it is necessary that it is a valid infra_id to write it in the db

I wonder if we even need to return a `TestFixture` in this case, since the delete of the infra should cascade the deletion of the objects referencing it

File: editoast/src/views/pathfinding/catenaries.rs
Comment (Castavo) [reply editoast/src/views/pathfinding/catenaries.rs:68]: With the --release flag, the endpoint takes between 15 and 20ms. I tried your suggested change and it did not affect the computation time
I propose I leave the HashSet because it makes fewer lines and it's more straight forward.

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: We can mark this as TODO (or remove it)

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: Do we have a ticket for fixing the import of catenaries? We should link this TODO to it.

```suggestion
                    // TODO: We should send a 400 here, but that would make this endpoint unusable until
```

**Review by flomonster — CHANGES_REQUESTED** [review]

**Review by leovalais — DISMISSED** [review]
LGTM, tested GET endpoint

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: We can define an enum `PathfindingError` and use the `editoast_error` macro.

File: editoast/src/error.rs
Comment (flomonster) [inline editoast/src/error.rs]:
```suggestion
#[derive(Debug, Serialize, Deserialize)]
```

**Review by flomonster — CHANGES_REQUESTED** [review]
Almost good for me.

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: Since `infra_id` is a foreign key will always find the infra. Meaning we can `expect` to have the infra.

**Review by flomonster — APPROVED** [review]
LGTM. I only have one suggestion left.

File: editoast/src/views/pathfinding/catenaries.rs
Comment (Castavo) [reply editoast/src/views/pathfinding/catenaries.rs]: Makes sense

File: editoast/src/views/pathfinding/catenaries.rs
Comment (flomonster) [inline editoast/src/views/pathfinding/catenaries.rs]: We can transform the `PathfindingError` into `InternalError` to have these warnings correctly serialized.

```suggestion
    warnings: Vec<InternalError>,
```

**Review by flomonster — APPROVED** [review]

File: editoast/src/views/pathfinding/catenaries.rs
Comment (Castavo) [reply editoast/src/views/pathfinding/catenaries.rs]: Done

**Pull request merged**
